### PR TITLE
Expose locations of `_ViewImports.cshtml` that affect a given Razor file.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Added `ChunkTreeResult` to associate inherited chunks with a specific source file.

#2256

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs [hunk 1]
```diff
@@ -224,6 +225,15 @@ internal ChunkInheritanceUtility ChunkInheritanceUtility
             }
         }
 
+        public IReadOnlyList<ChunkTreeResult> GetInheritedChunkTreeResults(string sourceFileName)
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs]: Would this need to be on the interface?

File: src/Microsoft.AspNet.Mvc.Razor.Host/Directives/ChunkTreeResult.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNet.Razor.Chunks;
+
+namespace Microsoft.AspNet.Mvc.Razor.Directives
+{
+    public class ChunkTreeResult
+    {
+        public ChunkTree ChunkTree { get; set; }
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor.Host/Directives/ChunkTreeResult.cs]: Getter only.

**Comment by pranavkm** [comment]
Design looks good.

File: src/Microsoft.AspNet.Mvc.Razor.Host/Directives/ChunkInheritanceUtility.cs [hunk 3]
```diff
@@ -40,17 +40,18 @@ public class ChunkInheritanceUtility
         }
 
         /// <summary>
-        /// Gets an ordered <see cref="IReadOnlyList{T}"/> of parsed <see cref="ChunkTree"/> for each
-        /// <c>_ViewImports</c> that is applicable to the page located at <paramref name="pagePath"/>. The list is
-        /// ordered so that the <see cref="ChunkTree"/> for the <c>_ViewImports</c> closest to the
+        /// Gets an ordered <see cref="IReadOnlyList{ChunkTreeResult}"/> of parsed <see cref="ChunkTree"/>s and
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor.Host/Directives/ChunkInheritanceUtility.cs]: FYI `<see cref="IReadOnlyList{T}"/>` and `<see cref="IReadOnlyList{ChunkTreeResult}"/>` are displayed the same. That's why these comments include bits like `of parsed <see cref="ChunkTree"/>`.

File: src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs (hunk 1 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs]: Nope, at least not today. It's consumed by tooling who knows of/constructs `MvcRazorHost`

File: src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs [hunk 4]
```diff
@@ -273,5 +279,14 @@ public override ParserBase DecorateCodeParser([NotNull] ParserBase incomingCodeP
                     CreateModelExpressionMethodName = CreateModelExpressionMethod
                 });
         }
+
+        private IReadOnlyList<ChunkTree> GetInheritedChunkTrees(string sourceFileName)
+        {
+            var inheritedChunkTrees = GetInheritedChunkTreeResults(sourceFileName)
+                .Select(result => result.ChunkTree)
+                .ToList();
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs]: Curious: Why do these methods deal in `IReadOnlyList<T>` versus `IEnumerable<T>`? Don't appear to have any cases where anything is done except iterating through the collections -- _once_.

Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs]: Just following what we did before. @pranavkm was there a specific reason for `IReadOnlyList` (I can't recall)?

**Comment by NTaylorMullen** [comment]
Updated. No longer design.

File: src/Microsoft.AspNet.Mvc.Razor.Host/Directives/ChunkTreeResult.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    /// </summary>
+    public class ChunkTreeResult
+    {
+        /// <summary>
+        /// Initializes a new instance of <see cref="ChunkTreeResult"/>.
+        /// </summary>
+        /// <param name="chunkTree">The <see cref="AspNet.Razor.Chunks.ChunkTree"/> generated from the file at the
+        /// given <paramref name="filePath"/>.</param>
+        /// <param name="filePath">The path to the file that generated the given <paramref name="chunkTree"/>.</param>
+        public ChunkTreeResult(ChunkTree chunkTree, string filePath)
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor.Host/Directives/ChunkTreeResult.cs]: `[NotNull]`

File: src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs (hunk 4 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs]: We rely on the ordering of chunks to inject them in the page being generated. The `IList` contract is meant to indicate this ordering.

File: test/Microsoft.AspNet.Mvc.Razor.Host.Test/Directives/ChunkInheritanceUtilityTest.cs [hunk 6]
```diff
@@ -36,14 +35,15 @@ @inherits MyBaseType
...
+                PlatformNormalizer.NormalizePath(@"Views\home\Index.cshtml"));
 
             // Assert
-            Assert.Collection(chunkTrees,
-                chunkTree =>
+            Assert.Collection(chunkTreeResults,
+                chunkTreeResult =>
                 {
                     var viewImportsPath = PlatformNormalizer.NormalizePath(@"Views\home\_ViewImports.cshtml");
-                    Assert.Collection(chunkTree.Chunks,
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Razor.Host.Test/Directives/ChunkInheritanceUtilityTest.cs]: Could you throw in an `Assert.Equal(viewImportsPath, chunkTreeResult.FilePath);` here?

File: src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs [hunk 7]
```diff
@@ -224,6 +225,21 @@ internal ChunkInheritanceUtility ChunkInheritanceUtility
             }
         }
 
+        /// <summary>
+        /// Locates and parses _ViewImports.cshtml files applying to the given <paramref name="sourceFileName"/> to
+        /// create <see cref="ChunkTreeResult"/>s.
+        /// </summary>
+        /// <param name="sourceFileName">The path to a Razor file to locate _ViewImports.cshtml for.</param>
+        /// <returns>Inherited <see cref="ChunkTreeResult"/>s.</returns>
+        public IReadOnlyList<ChunkTreeResult> GetInheritedChunkTreeResults(string sourceFileName)
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs]: `NotNull`

**Comment by pranavkm** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs (hunk 4 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs]: `IReadOnlyList` indicates the collection is enumerable and has an indexer and `Count` property. Only need `IEnumerable<T>` to indicate ordering. (Detail of the ordering e.g. the lack of documented order for entries in a `Dictionary` is a separate question.)

Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs]: ... Hopefully an API review will revisit these APIs and decide whether `IReadOnlyList<T>` is needed here.

File: src/Microsoft.AspNet.Mvc.Razor.Host/Directives/ChunkTreeResult.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+        /// <summary>
+        /// The <see cref="AspNet.Razor.Chunks.ChunkTree"/> generated from the file at <see cref="FilePath"/>.
+        /// </summary>
+        public ChunkTree ChunkTree { get; }
+
+        /// <summary>
+        /// The path to the file that generated the <see cref="ChunkTree"/>.
+        /// </summary>
+        public string FilePath { get; }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor.Host/Directives/ChunkTreeResult.cs]: No tests confirm the `FilePath` value of `ChunkTreeResult`s.

File: src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs [hunk 9]
```diff
@@ -224,6 +225,21 @@ internal ChunkInheritanceUtility ChunkInheritanceUtility
             }
         }
 
+        /// <summary>
+        /// Locates and parses _ViewImports.cshtml files applying to the given <paramref name="sourceFileName"/> to
+        /// create <see cref="ChunkTreeResult"/>s.
+        /// </summary>
+        /// <param name="sourceFileName">The path to a Razor file to locate _ViewImports.cshtml for.</param>
+        /// <returns>Inherited <see cref="ChunkTreeResult"/>s.</returns>
+        public IReadOnlyList<ChunkTreeResult> GetInheritedChunkTreeResults([NotNull] string sourceFileName)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor.Host/MvcRazorHost.cs]: This method is not tested.

**Comment by dougbu** [comment]
Merged but needs more tests.

**Pull request merged**